### ISSUE TITLE
Intermittent connectivity issues between pods on different nodes

### ISSUE THREAD

**Issue by Hayajiro** [cilium/cilium#23117]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We're experiencing intermittent issues with pod-to-pod traffic between nodes. For most of the time, connection works perfectly fine without any issues. However, at some point, the pods on some of our nodes start to experience connectivity issues. Those issues seen somewhat related to node reboots due to kernel upgrades, **but it's not always the case**, as we've also seen it happen in the middle of the night, already.

When the issue occurs, all communication between pods on the affected nodes dies completely. Directly connecting from Node 1, which is where pods can't seem to communicate with CoreDNS on Node 2, to the respective CoreDNS instance on Node 2 works just fine. This was first noticed with failing DNS-requests, which made it look like DNS / UDP issues at first. However, TCP (e.g. HTTP) traffic is also affected by this.

~~The first time we noticed the issues was after upgrading our clusters to Ubuntu 22.04 (from 20.04) and Cilium 1.12 (from 1.11). On the first cluster, we ran both upgrades at the same time, but after noticing the issue, we only upgraded our Ubuntu version to try and pin-point it. However, even with the older Cilium version, we experienced the same problems. This makes it look as if it might be an incompatibility with Ubuntu 22.04 (or, rather its Kernel).~~ Upon further inspection, the issue seems to have existed for about a year now, but it seems to have become more severe with the recent upgrades.

Some things we've tried so far (to no avail):

* Update Cilium (from 1.12.3 to 1.12.5), as we saw some changes that looked related to the issue we are experiencing
* Restart the Cilium-DaemonSet in an attempt to temporarily fix the issue in an affected cluster
* Ensure `rp_filter` is set to `0` on for `all` and all cilium-related interfaces (only `default`, `eth0` and `lo` are set to `2`)
* Ensure `systemd-networkd` is configured properly (`ManageForeignRoutes` and `ManageForeignRoutingPolicyRules` set to `no`)
* Ensure `/sys/fs/bpf` is already mounted before `clilium-agent` starts

The only thing that reliably manages to fix the issue seems to be restarting the affected node(s). Of course, that only helps until the issue resurfaces a few days / weeks later.

### Cilium Version

cilium-cli: v0.12.12 compiled with go1.19.4 on linux/amd64
cilium image (default): v1.12.5
cilium image (stable): v1.12.5
cilium image (running): v1.12.5

### Kernel Version

Linux m1.k8s-test-1 5.15.0-58-generic #64-Ubuntu SMP Thu Jan 5 11:43:13 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.9", GitCommit:"c1de2d70269039fe55efb98e737d9a29f9155246", GitTreeState:"clean", BuildDate:"2022-07-13T14:26:51Z", GoVersion:"go1.17.11", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.9", GitCommit:"c1de2d70269039fe55efb98e737d9a29f9155246", GitTreeState:"clean", BuildDate:"2022-07-13T14:19:57Z", GoVersion:"go1.17.11", Compiler:"gc", Platform:"linux/amd64"}

### Sysdump

Upon request, I can provide the complete sysdump via Slack or other means. Individual files / log excerpts can also be provided directly in this issue.

### Relevant log output

```shell
# Traefik-pod on node 2, able to communicate with CoreDNS (same node), but not with Prometheus (node 1):

/ $ wget -O- http://prometheus.prometheus.svc.cluster.local:9090
Connecting to prometheus.prometheus.svc.cluster.local:9090 (10.99.118.204:9090)
# Stuck with this

---

# Traefik-pod on node 1, which can't communicate with CoreDNS on node 2.

/ $ wget -O- http://prometheus.prometheus.svc.cluster.local:9090
wget: bad address 'prometheus.prometheus.svc.cluster.local:9090'

/ $ nslookup prometheus.prometheus.svc.cluster.local 10.244.9.67
;; connection timed out; no servers could be reached

---

# Same Traefik-pod on node 1, but connecting to CoreDNS on node 3, which works just fine:

/ $ nslookup prometheus.prometheus.svc.cluster.local 10.244.10.214
Server:		10.244.10.214
Address:	10.244.10.214:53


Name:	prometheus.prometheus.svc.cluster.local
Address: 10.99.118.204
```

### Anything else?

The issue was first noticed with our CI-Clusters, which constantly create new (short-lived) namespaces and pods for our pipelines and jobs, respectively. This made us believe that the issue might be related to those high-frequency updates. However, we started noticing those issues in our other clusters where pods are relatively static (and only see updates when new versions are deployed to their respective environments).

A few words on our setup:

* Cluster installed using kubeadm
* Cilium installed via helm
  * Using the following args:
```
  - "--namespace kube-system"
  - "--set ipam.mode=kubernetes"
  - "--set hubble.relay.enabled=true"
  - "--set hubble.ui.enabled=true"
  - "--set l7Proxy=false"
  - "--set encryption.enabled=true"
  - "--set encryption.type=wireguard"
  - "--set prometheus.enabled=true"
  - "--set operator.prometheus.enabled=true"
  - "--set hubble.metrics.enabled='{dns,drop,tcp,flow,icmp,http}'"
  - "--set hostFirewall.enabled=true"
  - "--set cni.chainingMode=portmap"
```
* Output of `cilium status`:
```
KVStore:                          Ok   Disabled
Kubernetes:                       Ok   1.23 (v1.23.9) [linux/amd64]
Kubernetes APIs:                  ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:             Probe     [eth0 XX.XX.XX.6]
Host firewall:                    Enabled   [eth0]
CNI Chaining:                     portmap
Cilium:                           Ok   1.12.5 (v1.12.5-701acde)
NodeMonitor:                      Listening for events on 4 CPUs with 64x4096 of shared memory
Cilium health daemon:             Ok
IPAM:                             IPv4: 14/254 allocated from 10.244.10.0/24,
BandwidthManager:                 Disabled
Host Routing:                     Legacy
Masquerading:                     BPF   [eth0]   10.244.10.0/24 [IPv4: Enabled, IPv6: Disabled]
Controller Status:                81/81 healthy
Proxy Status:                     No managed proxy redirect
Global Identity Range:            min 256, max 65535
Hubble:                           Ok              Current/Max Flows: 4095/4095 (100.00%), Flows/s: 36.11   Metrics: Ok
Encryption:                       Wireguard       [cilium_wg0 (Pubkey: [...], Port: 51871, Peers: 6)]
Cluster health:                   1/7 reachable   (2023-01-16T13:46:52Z)
  Name                        IP            Node          Endpoints
  m1.k8s-test-1.example.com   XX.XX.XX.1    unreachable   reachable
  m2.k8s-test-1.example.com   XX.XX.XX.2    unreachable   reachable
  m3.k8s-test-1.example.com   XX.XX.XX.3    unreachable   reachable
  n1.k8s-test-1.example.com   XX.XX.XX.4    unreachable   reachable
  n2.k8s-test-1.example.com   XX.XX.XX.5    unreachable   reachable
  n4.k8s-test-1.example.com   XX.XX.XX.7    unreachable   reachable
```
* Output of `cilium-health status` (same for all entries)
```
  m1.k8s-test-1.example.com:
    Host connectivity to XX.XX.XX.1:
      ICMP to stack:   Connection timed out
      HTTP to agent:   OK, RTT=1.398567ms
    Endpoint connectivity to 10.244.1.20:
      ICMP to stack:   OK, RTT=2.133171ms
      HTTP to agent:   OK, RTT=1.043445ms
```

Failing ICMP Host connectivity check is expected, as it's not allowed in our network policies.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by brb** [comment]
Thanks for the issue!

> When the issue occurs, all communication between pods on the affected nodes dies completely.

I think this could have happened because the node after the reboot has lost its WireGuard private key, and generated a new one. Could you check that both nodes are using the same public keys?

**Comment by Hayajiro** [comment]
> Could you check that both nodes are using the same public keys?

Yes, they all use the same public keys for the respective nodes. On both affected nodes, the last handshakes are all relatively recent (within 2 minutes) and all peers indicate both sent and received traffic, so the WireGuard tunnel seems functional.

**Comment by brb** [comment]
Then check whether you see any increasing drops in `cilium metrics list | grep drop`.

**Comment by Hayajiro** [comment]
There are drops related to `Policy denied` that are constantly increasing on node 2. I could swear that I've checked this before and there were no drops when the issue last occurred. When looking at the monitor, we can see the following drops:
```
xx drop (Policy denied) flow 0x0 to endpoint 0, file bpf_host.c line 665, , identity world->unknown: [node1]:55517 -> [node2]:37419 udp
xx drop (Policy denied) flow 0x0 to endpoint 0, file bpf_host.c line 665, , identity world->unknown: [master1]:40586 -> [node2]:49482 udp
```

Looks like one of our control-plane nodes also has the issue. However, those ports are not related to anything we're (expecting to be) running services on, `netstat` doesn't show anything for those ports, either.

But there's something: When checking the currently applied WireGuard config again, I see those same ports the two nodes are trying to connect to - as Peer port for node 2... So it seems that, while the WireGuard configuration correctly remembers its keys, n1 (as well as m1, which we just noticed) seems to not remember the Peer port of n2.

I guess this also explains why the tunnel **looked** functional when looking at the output of `wg`. One side of the tunnel (n2 -> n1) is properly configured, while the other side (n1 -> n2) uses the wrong port.

Any ideas what could cause Cilium / WireGuard to forget the correct peer port?

**Comment by Hayajiro** [comment]
> Are you sure that it's the WireGuard tunnel traffic?

Yes, running `wg` on n1 explicitly shows `37419` as the endpoint port for the peer that represents n1. For all other nodes, it correctly uses `51871`. This issue also only exists on n1 and m1 (each with different ports for n2), all other nodes and masters use the correct port for n2 (as well as all other peers).

**Comment by Hayajiro** [comment]
No, Cilium is the only thing configuring WireGuard. `wireguard-tools` isn't even installed by default, and the only WireGuard-related configuration is specifying the `encryption.type` in our Cilium deployment. So everything WireGuard-related should be managed solely by Cilium.

**Comment by brb** [comment]
Is it possible for you to run strace on the `cilium-agent` to see what it's actually doing during the WG initialization?

**Comment by Hayajiro** [comment]
I was able to strace the running `cilium-agent` on node 1 while restarting `cilium-agent` (which I hope triggers the WireGuard init) on node 2. However, there's so much going on (even before restarting the other `cilium-agent` instance), that it's hard to find anything useful (at least without knowing exactly what to look for). Looking for generic strings like `wg`, `wireguard`, `endpoint` etc. didn't yield anything that seems to be related.

However, after restarting `cilium-agent` on node 2, the issue is still present.

**Comment by Hayajiro** [comment]
Sure, here you go:
```
root@n1:/home/cilium# cilium status --verbose
KVStore:                Ok   Disabled
Kubernetes:             Ok   1.23 (v1.23.9) [linux/amd64]
Kubernetes APIs:        ["cilium/v2::CiliumClusterwideNetworkPolicy", "cilium/v2::CiliumEndpoint", "cilium/v2::CiliumNetworkPolicy", "cilium/v2::CiliumNode", "core/v1::Namespace", "core/v1::Node", "core/v1::Pods", "core/v1::Service", "discovery/v1::EndpointSlice", "networking.k8s.io/v1::NetworkPolicy"]
KubeProxyReplacement:   Probe     [eth0 $NODE1_IP]
Host firewall:          Enabled   [eth0]
CNI Chaining:           portmap
Cilium:                 Ok   1.12.5 (v1.12.5-701acde)
NodeMonitor:            Listening for events on 4 CPUs with 64x4096 of shared memory
Cilium health daemon:   Ok
IPAM:                   IPv4: 14/254 allocated from 10.244.8.0/24,
Allocated addresses:
  10.244.8.102 (router)
  10.244.8.104 (dex/dex-67647c7d47-dgm4k[restored])
  10.244.8.12 (logging/fluent-bit-5j44r[restored])
  10.244.8.121 (cert-manager/cert-manager-webhook-785bb86798-p8gx8[restored])
  10.244.8.134 (traefik/traefik-xtd7h[restored])
  10.244.8.164 (flux-system/helm-controller-59dcbc6dcb-vscmh[restored])
  10.244.8.173 (prometheus/prometheus-5666b7b8f7-4cbl5[restored])
  10.244.8.188 (kube-system/kured-w84zz[restored])
  10.244.8.238 (diagrams-test-1-hw-node-3436-production/diagrams-855f9f8cf8-4mx4c[restored])
  10.244.8.27 (kube-system/hubble-ui-77b4dfdf8f-6h4kl[restored])
  10.244.8.57 (prometheus/prometheus-kube-state-metrics-784497b785-qj6mw[restored])
  10.244.8.72 (kube-system/hcloud-csi-node-lw9fq[restored])
  10.244.8.78 (health)
  10.244.8.91 (dex/dex-k8s-authenticator-7b78dd44b-nxv4x[restored])
BandwidthManager:       Disabled
Host Routing:           Legacy
Masquerading:           BPF   [eth0]   10.244.8.0/24 [IPv4: Enabled, IPv6: Disabled]
Clock Source for BPF:   ktime
Controller Status:      81/81 healthy
  Name                                          Last success    Last error      Count   Message
  bpf-map-sync-cilium_lxc                       7s ago          never           0       no error
  cilium-health-ep                              42s ago         never           0       no error
  dns-garbage-collector-job                     15s ago         never           0       no error
  endpoint-1221-regeneration-recovery           never           never           0       no error
  endpoint-1239-regeneration-recovery           never           never           0       no error
  endpoint-1976-regeneration-recovery           never           never           0       no error
  endpoint-2148-regeneration-recovery           never           never           0       no error
  endpoint-232-regeneration-recovery            never           never           0       no error
  endpoint-239-regeneration-recovery            never           never           0       no error
  endpoint-2671-regeneration-recovery           never           never           0       no error
  endpoint-368-regeneration-recovery            never           never           0       no error
  endpoint-369-regeneration-recovery            never           never           0       no error
  endpoint-3900-regeneration-recovery           never           never           0       no error
  endpoint-392-regeneration-recovery            never           never           0       no error
  endpoint-431-regeneration-recovery            never           never           0       no error
  endpoint-747-regeneration-recovery            never           never           0       no error
  endpoint-902-regeneration-recovery            never           never           0       no error
  endpoint-gc                                   1m16s ago       never           0       no error
  ipcache-inject-labels                         28h31m1s ago    51h16m17s ago   0       no error
  k8s-heartbeat                                 26s ago         never           0       no error
  link-cache                                    10s ago         never           0       no error
  metricsmap-bpf-prom-sync                      7s ago          never           0       no error
  resolve-identity-1239                         1m28s ago       never           0       no error
  resolve-identity-3900                         1m28s ago       never           0       no error
  resolve-identity-902                          50s ago         never           0       no error
  restoring-ep-identity (1221)                  51h15m51s ago   never           0       no error
  restoring-ep-identity (1239)                  51h15m51s ago   never           0       no error
  restoring-ep-identity (1976)                  51h15m51s ago   never           0       no error
  restoring-ep-identity (2148)                  51h15m51s ago   never           0       no error
  restoring-ep-identity (232)                   51h15m51s ago   never           0       no error
  restoring-ep-identity (239)                   51h15m51s ago   never           0       no error
  restoring-ep-identity (2671)                  51h15m51s ago   never           0       no error
  restoring-ep-identity (368)                   51h15m51s ago   never           0       no error
  restoring-ep-identity (369)                   51h15m51s ago   never           0       no error
  restoring-ep-identity (3900)                  51h15m51s ago   never           0       no error
  restoring-ep-identity (392)                   51h15m51s ago   never           0       no error
  restoring-ep-identity (431)                   51h15m51s ago   never           0       no error
  restoring-ep-identity (747)                   51h15m51s ago   never           0       no error
  sync-endpoints-and-host-ips                   41s ago         never           0       no error
  sync-lb-maps-with-k8s-services                51h15m51s ago   never           0       no error
  sync-policymap-1221                           36s ago         never           0       no error
  sync-policymap-1239                           36s ago         never           0       no error
  sync-policymap-1976                           36s ago         never           0       no error
  sync-policymap-2148                           36s ago         never           0       no error
  sync-policymap-232                            36s ago         never           0       no error
  sync-policymap-239                            36s ago         never           0       no error
  sync-policymap-2671                           36s ago         never           0       no error
  sync-policymap-368                            36s ago         never           0       no error
  sync-policymap-369                            36s ago         never           0       no error
  sync-policymap-3900                           36s ago         never           0       no error
  sync-policymap-392                            36s ago         never           0       no error
  sync-policymap-431                            36s ago         never           0       no error
  sync-policymap-747                            36s ago         never           0       no error
  sync-policymap-902                            36s ago         never           0       no error
  sync-to-k8s-ciliumendpoint (1221)             4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (1239)             3s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (1976)             4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (2148)             4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (232)              4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (239)              4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (2671)             4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (368)              4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (369)              5s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (3900)             3s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (392)              4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (431)              4s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (747)              8s ago          never           0       no error
  sync-to-k8s-ciliumendpoint (902)              5s ago          never           0       no error
  template-dir-watcher                          never           never           0       no error
  waiting-initial-global-identities-ep (1221)   51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (1239)   51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (1976)   51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (2148)   51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (232)    51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (239)    51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (2671)   51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (368)    51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (3900)   51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (392)    51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (431)    51h15m51s ago   never           0       no error
  waiting-initial-global-identities-ep (747)    51h15m51s ago   never           0       no error
Proxy Status:            No managed proxy redirect
Global Identity Range:   min 256, max 65535
Hubble:                  Ok   Current/Max Flows: 4095/4095 (100.00%), Flows/s: 42.98   Metrics: Ok
KubeProxyReplacement Details:
  Status:                 Probe
  Socket LB:              Enabled
  Socket LB Protocols:    TCP, UDP
  Devices:                eth0 $NODE1_IP
  Mode:                   SNAT
  Backend Selection:      Random
  Session Affinity:       Enabled
  Graceful Termination:   Enabled
  NAT46/64 Support:       Disabled
  XDP Acceleration:       Disabled
  Services:
  - ClusterIP:      Enabled
  - NodePort:       Enabled (Range: 30000-32767)
  - LoadBalancer:   Enabled
  - externalIPs:    Enabled
  - HostPort:       Enabled
BPF Maps:   dynamic sizing: on (ratio: 0.002500)
  Name                          Size
  Non-TCP connection tracking   71852
  TCP connection tracking       143705
  Endpoint policy               65535
  Events                        4
  IP cache                      512000
  IP masquerading agent         16384
  IPv4 fragmentation            8192
  IPv4 service                  65536
  IPv6 service                  65536
  IPv4 service backend          65536
  IPv6 service backend          65536
  IPv4 service reverse NAT      65536
  IPv6 service reverse NAT      65536
  Metrics                       1024
  NAT                           143705
  Neighbor table                143705
  Global policy                 16384
  Per endpoint policy           65536
  Session affinity              65536
  Signal                        4
  Sockmap                       65535
  Sock reverse NAT              71852
  Tunnel                        65536
Encryption:                                   Wireguard       [cilium_wg0 (Pubkey: [....], Port: 51871, Peers: 6)]
Cluster health:                               1/7 reachable   (2023-01-18T15:30:52Z)
  Name                                    IP              Node          Endpoints
  n1.k8s-test-1.example.com (localhost)   $NODE1_IP       reachable     reachable
  m1.k8s-test-1.example.com               $MASTER1_IP     unreachable   reachable
  m2.k8s-test-1.example.com               $MASTER2_IP     unreachable   reachable
  m3.k8s-test-1.example.com               $MASTER3_IP     unreachable   reachable
  n2.k8s-test-1.example.com               $NODE2_IP       unreachable   reachable
  n3.k8s-test-1.example.com               $NODE3_IP       unreachable   reachable
  n4.k8s-test-1.example.com               $NODE4_IP       unreachable   reachable
```

```
root@n1:/home/cilium# cilium bpf ipcache get $NODE2_IP
$NODE2_IP maps to identity identity=6 encryptkey=0 tunnelendpoint=$NODE2_IP
```

**Comment by brb** [comment]
Thanks. Do you also have `cilium bpf ipcache get $NODE1_IP`?

**Comment by Hayajiro** [comment]
> Do you also have `cilium bpf ipcache get $NODE1_IP`?

Sorry, I was out of office already. Executed from node 1, it looks like this:
```
root@n1:/home/cilium# cilium bpf ipcache get $NODE1_IP
$NODE1_IP maps to identity identity=1 encryptkey=0 tunnelendpoint=0.0.0.0
```

On node 2, it looks like this:
```
root@n2:/home/cilium# cilium bpf ipcache get $NODE1_IP
$NODE1_IP maps to identity identity=6 encryptkey=0 tunnelendpoint=$NODE1_IP
```

**Comment by brb** [comment]
Did you execute those cmds when you were experiencing the problem?

**Comment by Hayajiro** [comment]
Yes, those were executed while WireGuard was using the incorrect port for the connection between the respective nodes.

On our productive clusters, it seems (for the last two days since the changes have been gradually deployed) like it was fixed by setting `kubeProxyReplacement` to `disabled` from the previous default of `probe`, which I noticed was deprecated in 1.12.  
The behavior of `probe` would probably also explain why the issues got worse with the upgrade to Ubuntu 22.04, given that it uses a much more recent kernel.

I guess Cilium sometimes confuses the WireGuard traffic for cluster-internal traffic and proceeds to SNAT it?

**Comment by mitchese** [comment]
We've just hit this issue in our cluster as well. What happened: a node was suddenly unreachable with no reason/no change on our side. We were alarmed to all services on a node going offline/unreachable.

We found inspecting the wireguard configuration with wg showconf cilium_wg0, the Endpoint was erroneously updating to port 49590.  The entire cluster is running on 51871, except for host2 has learned that when talking to host1 it should use 49590.

host1:51871 -> host2:51871
host2:51871 -> host1:49590

All communication between host1 and host2 are broken, because host2 thinks it should reach back to host1 on udp49590.  Looking at the wireguard config, host1 has the private key/service on 51871 as expected and transmits to host2 on 51871.

host2 has learned that the endpoint for host1 is unexpectedly 49590.

we ran cilium bpf nat list on host1 and found there was a NAT entry for this connection
`UDP OUT 10.20.x.1:51871 -> 10.20.x.2:51871 XLATE_SRC 10.20.x.1:49590 Created=363645sec HostLocal=1`

We believe this NAT entry is the source of losing the connectivity to the node. While wireguard is sending the traffic out on port 51871 destined for host2, the NAT is translating this to 49590 before it goes out the physical interface, and therefore worker2 learns the endpoint is 49590.
